src/librustc_ast/util/comments.rs

```diff
@@ -28,30 +30,43 @@ pub struct Comment {
     pub pos: BytePos,
 }
 
+/// Check if `s` a line doc comment.
+///
+/// Like "/// [...]" or "//! [...]" but not "//// [...]".
 pub fn is_line_doc_comment(s: &str) -> bool {
-    let res = (s.starts_with("///") && *s.as_bytes().get(3).unwrap_or(&b' ') != b'/')
-        || s.starts_with("//!");
-    debug!("is {:?} a doc comment? {}", s, res);
-    res
+    let yes = match s.as_bytes() {
+        [b'/', b'/', b'/', b'/', ..] => false,
+        [b'/', b'/', b'/', ..] => true,
+        [b'/', b'/', b'!', ..] => true,
+        _ => false,
+    };
+    debug!("is {:?} a line doc comment? {}", s, yes);
+    yes
 }
 
+/// Check if `s` a block doc comment.
+///
+/// Like "/** [...] */" or "/*! [...] */" but not "/*** [...] */".
 pub fn is_block_doc_comment(s: &str) -> bool {
-    // Prevent `/**/` from being parsed as a doc comment
-    let res = ((s.starts_with("/**") && *s.as_bytes().get(3).unwrap_or(&b' ') != b'*')
-        || s.starts_with("/*!"))
-        && s.len() >= 5;
-    debug!("is {:?} a doc comment? {}", s, res);
-    res
+    // Previously, `/**/` was incorrectly regarded as a doc comment because it
+    // starts with `/**` and ends with `*/`. However, this caused an ICE
+    // because some code assumed that the length of a doc comment is at least 5.
+    let yes = match s.as_bytes() {
+        [b'/', b'*', b'*', b'*', _, ..] => false,
+        [b'/', b'*', b'*', _, _, ..] => true,
+        [b'/', b'*', b'!', _, _, ..] => true,
+        _ => false,
```

Comment:
Thinking a bit more, I think I should just match on `b"/**/" => false`.
```suggestion
    // because some code assumed that the length of a doc comment is at least 5.
    let yes = match s.as_bytes() {
        [b'/', b'*', b'*', b'*', ..] => false,
        [b'/', b'*', b'*', b'/', ..] => false,
        [b'/', b'*', b'*', ..] => true,
        [b'/', b'*', b'!', ..] => true,
        _ => false,
```